ftms_filtering.qmd

```diff
@@ -119,179 +126,195 @@ perc_ms2_assigned <- (n_assigned_ms2 / n_total_ms2) * 100
 
 ```{r}
 select_unique_MS2 <- function(ms2, ftms) {
-  
+
   sp <- spectraData(ms2)
-  
+
   # Create a unique identifier for each unique groupe of (dataOrigin, scanIndex)
   sp$groups <- paste0(sp$dataOrigin, "_", sp$scanIndex)
-  
-  # Create indices for duplicated groups 
+
+  # Create indices for duplicated groups
   group_idx <- split(seq_len(nrow(sp)), sp$groups)
-  
+
+  # define MS2 as ambiguous if they are duplicated in sp (i.e., are assigned
+  # to >= 2 features. We use the unique ID for the MS2 spectra and their
+  # indices in sp for that.
+  sp$is_ambiguous <- FALSE
+  sp$is_ambiguous[sp$groups %in% names(group_idx)[lengths(group_idx) > 1]] <- TRUE
+
   # Initialization
   final_feature <- vector("list", nrow(sp))  # For the chosen features
   amb_f <- vector("list", nrow(sp))          # For ambiguous features
-  
+
   # Table of occurences for each feature in a dataOrigine
-  feature_table <- xtabs(~ feature_id + dataOrigin, data = sp)
-  
+  # jo: I think it's not good to use this table. it can happen (and it's not
+  # a problem) if we have 2 MS2 spectra for the same feature in the same
+  # sample (dataOrigin). The problem is if the same MS2 is assigned to
+  # two different features.
+  # feature_table <- xtabs(~ feature_id + dataOrigin, data = sp)
+
   # Loop on each group
   for (gname in names(group_idx)) {
     # line indices of each group
-    gidx <- group_idx[[gname]]  
+    gidx <- group_idx[[gname]]
     # all features that were assigned to the current group
-    feature_ids <- unique(unlist(sp$feature_id[gidx]))  
-    
+    feature_ids <- unique(unlist(sp$feature_id[gidx]))
+
     # if a scanIndex was assigned to a unique feature, we don't do anything
     if (length(feature_ids) == 1) {
       final_feature[[gidx[1]]] <- feature_ids
       amb_f[[gidx[1]]] <- character(0)
       next  # go to the next group...
     }
-    
+
     #rt of the first group
     ms2_rt <- rtime(ms2)[gidx[1]]
-    
-    
+
+
     # rtmed and intensity of the feature
     rt_features <- featureDefinitions(ftms)[feature_ids, "rtmed"]
 
     # Initialisation
     keep_feat <- character(0) #features to keep
     amb_feat <- feature_ids #ambiguous features
-    
+
     # Loop on each feature in the group
     for (i in seq_along(feature_ids)) {
       i_id <- feature_ids[i]
-      
-      # Case A 
-      # Verify if a feature was assigned at least once to a unique MS2 so it's
-      #an unambiguous feature
-      unamb_f <- any(feature_table[i_id, ] == 1)
-      if (unamb_f) {
-        # If yes we remove it's assignation to this scanIndex and 
-        #we marque other features as ambiguous
+
+      # Case A
+      # Verify if a feature was assigned at least once to a unique MS2 so there
+      # is an(other) unambiguous MS2-feature assignment. We thus skip/drop
+      # the current assignment and move to the next.
+      # unamb_f <- any(feature_table[i_id, ] == 1)
+      if (any(!sp$is_ambiguous[sp$feature_id == i_id])) {
```

Comment:
this is the new condition: if for the current feature there is any assignment to another MS2 which is **not ambiguous**, skip the current assignment and move to the next.